### PR TITLE
Create HierarchialClustering.md

### LINKED ISSUE #1335: [Idea]: Hierarchical Clustering Algorithm
### Idea Title

Implement Hierarchical Clustering Algorithm

### Idea Description

Hierarchical clustering is an unsupervised machine learning technique used to group similar data points into clusters based on their distance. This algorithm can be implemented using either an agglomerative (bottom-up) or divisive (top-down) approach. The agglomerative method starts with each data point as its own cluster and merges them based on a distance metric, while the divisive method begins with one cluster and splits it into smaller clusters. This idea aims to provide a robust hierarchical clustering algorithm with flexibility in distance measures (e.g., Euclidean, Manhattan) and linkage criteria (e.g., single, complete, average). The output will be a dendrogram representing the relationships between clusters, allowing users to visualize how clusters are formed.

### Potential Benefits

- [X] Improves user experience
- [X] Enhances learning
- [X] Adds a new feature
- [X] Improves code quality
- [X] Other

### Implementation Suggestions (Optional)

_No response_

### PR DESCRIPTION
**Pull request by ananyag309**
### Description
Added hierarchial clustering algorithm
Fixes #1335 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/machine-learning/HierarchialClustering.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
+# Hierarchical Clustering Algorithm
```
Comment (ajay-dhangar) [inline docs/machine-learning/HierarchialClustering.md:11]: please follow our format and syntax and also, carefully read our contribution guideline

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on our suggestion

**Comment by ananyag309** [comment]
@ajay-dhangar please review and merge it

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**